bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java

```diff
@@ -96,13 +96,38 @@
             this.ensemble = ensemble;
             this.eId = eId;
 
+            long entryIdToConsiderForWriteSet = eId;
+
+            if (clientCtx.getConf().enableStickyReads
+                    && lh.getLedgerMetadata().getWriteQuorumSize() == lh.getLedgerMetadata().getEnsembleSize()) {
```

Comment:
WQ == ES is quite limiting requirement. I.e. we run with ES = 7 and WQ = 3.
I'd suggest picking bookie with ensemble.id % WQ == 0.
i.e in case of 7 bookies in the ensemble and WQ 3, regular round-robin placement:

entry 0 is on bookies 0,1,2 (from ensemble) - reads stick to 0 (0 % 3 == 0)
entry 1 is on bookies 1,2,3 - reads stick to 3
entry 2 on 2,3,4 - reads sticks to 3
entry 3 on 3,4,5 - stick to 3
entry 4 on 4,5,6 - stick to 6
entry 5 on 5,6,7 - stick to 6
entry 6 on 6,7,0 - stick to 6 (choice of 6 and 0, got to the first one in order)
entry 7 on 7,0,1 - stick to 0.
etc

with WQ == ES it will stick to 0 all the time.

Plus it has to support read reordering (as I understand, it did not change) - don't try to use bookie if its netty channel is not writable.

Reply:
I thought of it but I think the IO reads saving will be just a fraction compared to E=W case. 

Eg, in your example, Bookie-0 will read e-0 and prefetch e-5 and e-6 though these 2 entries will be read on bookie 6

Reply:
but OS page cache will load more than e5 and e6 so all and all it will help.
I mean, it is speculations without perf tests anyways.

Reply:
Sure, but reading of e-5 and e-6 will be anyway "wasted".

Reply:
From strictly theoretical POV:
It works the same as picking bookie 0 for ES == WQ case.
For ES > WQ case it pins reads to limited subset of bookies and yes, something will get wasted but what else we can do other than round robin? 
in a full sequential read of ledger will this waste more or less than not handling ES > WQ case?

From practical POV, even with ES=WQ something will get wasted.
There is no guarantee that entries strictly sequential in the file. 1st, they can interleave with other ledgers.
Even with entry log per ledger *and* sorted ledger we can end up with out of order writes and out of order entries in entry log.

also, when you say prefetched do you mean OS file system cache or bufferedchannel in the bookie?